nx/include/switch/runtime/diag.h

```diff
@@ -0,0 +1,407 @@
+}
+
+NX_CONSTEXPR size_t diagComputeLogPacketPayloadSize(const DiagLogPacket *packet) {
+    return diagComputeChunkKeySize(&packet->payload.log_session_begin.header) +
+    diagComputeChunkKeySize(&packet->payload.log_session_end.header) +
+    diagComputeChunkKeySize(&packet->payload.text_log.header) +
+    diagComputeChunkKeySize(&packet->payload.line_number.header) +
+    diagComputeChunkKeySize(&packet->payload.file_name.header) +
+    diagComputeChunkKeySize(&packet->payload.function_name.header) +
+    diagComputeChunkKeySize(&packet->payload.module_name.header) +
+    diagComputeChunkKeySize(&packet->payload.thread_name.header) +
+    diagComputeChunkKeySize(&packet->payload.log_packet_drop_count.header) +
+    diagComputeChunkKeySize(&packet->payload.user_system_clock.header) +
+    diagComputeChunkKeySize(&packet->payload.process_name.header);
+}
+
+// Log payload encoding helpers
+
+NX_CONSTEXPR u8 *diagLogPayloadEncode(u8 *payload_buf, const void *data, size_t size) {
+    if(size > 0) {
+        __builtin_memcpy(payload_buf, data, size);
+        return payload_buf + size;
+    }
+    return payload_buf;
+}
+
+NX_CONSTEXPR u8 *diagLogPayloadEncodeU8ChunkType(u8 *payload_buf, DiagU8ChunkType *chunk_type) {
+    if(diagIsChunkTypeEmpty(&chunk_type->header)) {
+        return payload_buf;
+    }
+    u8 *buf = diagLogPayloadEncode(payload_buf, &chunk_type->header, sizeof(DiagLogDataChunkTypeHeader));
+    // Copy value.
+    return diagLogPayloadEncode(buf, &chunk_type->value, chunk_type->header.chunk_len);
+}
+
+NX_CONSTEXPR u8 *diagLogPayloadEncodeU32ChunkType(u8 *payload_buf, DiagU32ChunkType *chunk_type) {
+    if(diagIsChunkTypeEmpty(&chunk_type->header)) {
+        return payload_buf;
+    }
+    u8 *buf = diagLogPayloadEncode(payload_buf, &chunk_type->header, sizeof(DiagLogDataChunkTypeHeader));
+    // Copy value.
+    return diagLogPayloadEncode(buf, &chunk_type->value, chunk_type->header.chunk_len);
+}
+
+NX_CONSTEXPR u8 *diagLogPayloadEncodeU64ChunkType(u8 *payload_buf, DiagU64ChunkType *chunk_type) {
+    if(diagIsChunkTypeEmpty(&chunk_type->header)) {
+        return payload_buf;
+    }
+    u8 *buf = diagLogPayloadEncode(payload_buf, &chunk_type->header, sizeof(DiagLogDataChunkTypeHeader));
+    // Copy value.
+    return diagLogPayloadEncode(buf, &chunk_type->value, chunk_type->header.chunk_len);
+}
+
+NX_CONSTEXPR u8 *diagLogPayloadEncodeStringChunkType(u8 *payload_buf, DiagStringChunkType *chunk_type) {
+    if(diagIsChunkTypeEmpty(&chunk_type->header)) {
+        return payload_buf;
+    }
+    u8 *buf = diagLogPayloadEncode(payload_buf, &chunk_type->header, sizeof(DiagLogDataChunkTypeHeader));
+    // Copy string by its length.
+    return diagLogPayloadEncode(buf, chunk_type->value, chunk_type->header.chunk_len);
+}
+
+/**
+ * @brief Logs via lm.
+ * @note Retail lm is stubbed.
+ * @param[in] metadata Log parameters.
+ */
+NX_INLINE void diagLogImpl(const DiagLogMetadata *metadata) {
```

Comment:
Why is this an inline function? It should really be in a .c file...

Reply:
I wasn't sure whether to make it inline or place it in a .c file, and I decided to leave all diag code inline in case that would be better for performance (sdk doesn't inline it tho)

Reply:
Currently it looks absolutely terrible for code size :p

Reply:
What about U8/32/64/String chunk types and their helper functions?
Those should also be moved to a source file?

Reply:
All of these types look to me like internal implementation details, so yes

Reply:
Alright, then my idea is to leave Metadata and SourceInfo types since those are the types which contain log parametetrs, plus LogSeverity, and the rest would go inside diag.c as internal types.

Reply:
That's fine with me

Reply:
Another thing I'm considering is guarding the LogImpl function with a mutex, to guarantee that logging from different threads works fine

Reply:
Just did a quick test, and multithreaded logging works fine with a mutex guarding the function